modules/ding_webtrekk/ding_webtrekk.module

```diff
@@ -255,6 +258,42 @@ function ding_webtrekk_ding_entity_view($entity, $view_mode) {
   }
 }
 
+/**
+ * Implements hook_preprocess_node().
+ */
+function ding_webtrekk_preprocess_node(&$variables) {
+  $node = $variables['node'];
+  if ($node->type == 'ding_campaign_plus') {
+    // Create event that tracks when a campaign is viewed sending the title and
+    // internal ID.
+    $event_data = ding_webtrekk_create_event_data('Kampagne', [
+      DING_WEBTREKK_PARAMETER_CAMPAIGN_PLUS_TITLE => check_plain($node->title),
+      DING_WEBTREKK_PARAMETER_CAMPAIGN_PLUS_ID => $node->nid,
+    ]);
+
+    // We can't use the cacth-all event handler for this, since it relies on
+    // being triggered by a user interaction. Instead we send event when
+    // behavoirs are attached in frontend.
+    $event_data = json_encode($event_data);
+    $variables['attributes_array']['data-ding-webtrekk-event'] = $event_data;
+  }
+}
+
+/**
+ * Implements hook_process_node().
+ */
+function ding_webtrekk_process_node(&$variables) {
+  $node = $variables['node'];
+  if ($node->type == 'ding_campaign_plus') {
+    // Add u_navigatedby URL-parameter to campaign URLs, to be able to track
+    // page views triggered from campaigns. We use a process for this because
+    // campaign plus adds the variable in preprocess.
+    $url = &$variables['campaign_url'];
+    $url .= parse_url($url, PHP_URL_QUERY) !== NULL ? '&' : '?';
+    $url .= 'u_navigatedby=' . check_plain(rawurlencode($node->title));
```

Comment:
Instead of relying on string manipulation you could do the following using [url api functions](https://api.drupal.org/api/drupal/includes%21common.inc/function/drupal_parse_url/7.x):

```php
$url = drupal_parse_url($variables['campaign_url']);
$url['query']['u_navigatedby'] = $node->title;
$variables['campaign_url'] = url(
  $url['path'],
  [
    'query' => $url['query'],
    'fragment' => $url['fragment']
  ]
);
```

Reply:
Yep, this is the way to do it :) I see now that my code will also make a wrong URL if the original URL has fragment. Will fix!

Reply:
Fixed

Reply:
:+1: